storage/store/headers.go

```diff
@@ -21,34 +21,83 @@ type Headers struct {
 
 // NewHeaders creates a Headers instance, which stores block headers.
-// It supports storing, caching and retrieving by block ID and the additionally indexed by header height.
-func NewHeaders(collector module.CacheMetrics, db storage.DB) *Headers {
-	storeWithLock := func(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, header *flow.Header) error {
-		return operation.InsertHeader(lctx, rw, blockID, header)
-	}
-
-	retrieve := func(r storage.Reader, blockID flow.Identifier) (*flow.Header, error) {
-		var header flow.Header
-		err := operation.RetrieveHeader(r, blockID, &header)
-		return &header, err
+// It supports storing, caching and retrieving by block ID, and additionally indexes by header height and view.
+func NewHeaders(collector module.CacheMetrics, db storage.DB, chainID flow.ChainID) *Headers {
+	if chainID.IsClusterChain() {
+		panic("NewHeaders called on cluster chain ID - use NewClusterHeaders instead")
 	}
-
 	retrieveHeight := func(r storage.Reader, height uint64) (flow.Identifier, error) {
 		var id flow.Identifier
 		err := operation.LookupBlockHeight(r, height, &id)
 		return id, err
 	}
-
 	retrieveView := func(r storage.Reader, view uint64) (flow.Identifier, error) {
 		var id flow.Identifier
 		err := operation.LookupCertifiedBlockByView(r, view, &id)
 		return id, err
 	}
+	return newHeaders(collector, db, chainID, retrieveHeight, retrieveView)
+}
+
+// NewClusterHeaders creates a Headers instance for a collection cluster chain, which stores block headers for cluster blocks.
+// It supports storing, caching and retrieving by block ID, and additionally an index by header height.
+func NewClusterHeaders(collector module.CacheMetrics, db storage.DB, chainID flow.ChainID) *Headers {
+	if !chainID.IsClusterChain() {
+		panic("NewClusterHeaders called on non-cluster chain ID - use NewHeaders instead")
+	}
+	retrieveHeight := func(r storage.Reader, height uint64) (flow.Identifier, error) {
+		var id flow.Identifier
+		err := operation.LookupClusterBlockHeight(r, chainID, height, &id)
+		return id, err
+	}
+	retrieveView := func(r storage.Reader, height uint64) (flow.Identifier, error) {
+		var id flow.Identifier
+		return id, fmt.Errorf("retrieve by view not implemented for cluster headers")
+	}
+	return newHeaders(collector, db, chainID, retrieveHeight, retrieveView)
+}
+
+// newHeaders contains shared logic for Header storage, including storing and retrieving by block ID
+func newHeaders(collector module.CacheMetrics,
+	db storage.DB,
+	chainID flow.ChainID,
+	retrieveHeight retrieveFunc[uint64, flow.Identifier],
+	retrieveView retrieveFunc[uint64, flow.Identifier],
+) *Headers {
+	storeWithLock := func(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, header *flow.Header) error {
+		if header.ChainID != chainID {
+			return fmt.Errorf("expected chain ID %v, got %v: %w", chainID, header.ChainID, storage.ErrWrongChain)
+		}
+		if chainID.IsClusterChain() {
```

Comment:
Why do we need this check? From what I see, `operation.InsertHeader` already ensures that we hold a lock.

Reply:
Following up on [this discussion](https://github.com/onflow/flow-go/pull/8222#discussion_r2618797846): 

>  we can be more granular about the specific lock required

I agree. Though, I think this finer granularity should live in the lowest level possible. Thereby, we guarantee that all code paths must go through the fine-grained check.

I would suggest to differentiate on the level of the storage methods: i.e. split [`InsertHeader`](https://github.com/onflow/flow-go/blob/eefea22848d2f141f8e681dfc3c2e365f5374fb0/storage/operation/headers.go#L25) into two more specialized methods.